codec-redis/src/main/java/io/netty/handler/codec/redis/RedisDecoder.java

```diff
@@ -0,0 +1,232 @@
+            do {
+                switch (state) {
+                case DECODE_TYPE:
+                    next = decodeType(in);
+                    break;
+                case DECODE_INLINE:
+                    next = decodeInline(in, out);
+                    break;
+                case DECODE_LENGTH:
+                    next = decodeLength(in, out);
+                    break;
+                case DECODE_BULK_STRING:
+                    next = decodeBulkString(in, out);
+                    break;
+                default:
+                    throw new CodecException("Unknown state: " + state);
+                }
+            } while (next);
+        } catch (Exception e) {
+            resetDecoder();
+            throw new CodecException(e);
+        }
+    }
+
+    private void resetDecoder() {
+        state = State.DECODE_TYPE;
+    }
+
+    private boolean decodeType(ByteBuf in) throws Exception {
+        if (in.readableBytes() < 1) {
+            return false;
+        }
+        type = RedisMessageType.valueOf(in.readByte());
+        state = type.isInline() ? State.DECODE_INLINE : State.DECODE_LENGTH;
+        return true;
+    }
+
+    private boolean decodeInline(ByteBuf in, List<Object> out) throws Exception {
+        ByteBuf lineBytes = readLine(in);
+        if (lineBytes == null) {
+            return false;
+        }
+        addMessageToOut(out, newInlineRedisMessage(type, lineBytes));
+        return true;
+    }
+
+    private boolean decodeLength(ByteBuf in, List<Object> out) throws Exception {
+        ByteBuf lineByteBuf = readLine(in);
+        if (lineByteBuf == null) {
+            return false;
+        }
+        final long length = parseRedisNumber(lineByteBuf);
+        if (type == RedisMessageType.ARRAY_HEADER) {
+            addMessageToOut(out, RedisMessageFactory.createArrayHeader(length));
+            return true;
+        } else if (type == RedisMessageType.BULK_STRING) {
+            bulkStringLength = length;
+            state = State.DECODE_BULK_STRING;
+            return true;
+        } else {
+            throw new CodecException("bad type: " + type);
+        }
+    }
+
+    private boolean decodeBulkString(ByteBuf in, List<Object> out) throws Exception {
+        if (bulkStringLength == NULL_RESPONSE) {
+            // $-1\r\n
+            addMessageToOut(out, RedisMessageFactory.nullBulkString());
+        } else if (bulkStringLength == 0L) {
+            // $0\r\n <here> \r\n
+            if (in.readableBytes() < CRLF_LENGTH) {
+                return false;
+            }
+            in.skipBytes(CRLF_LENGTH);
+            addMessageToOut(out, RedisMessageFactory.emptyBulkString());
+        } else if (bulkStringLength > 0L) {
+            // ${bulkStringLength}\r\n <here> {data...}\r\n
+            if (in.readableBytes() < bulkStringLength + CRLF_LENGTH) {
+                return false;
+            }
+            ByteBuf content = in.readSlice((int) bulkStringLength);
```

Comment:
casting to `int` will truncate and may not protect against overflow. Should we handle overflow more gracefully?

Is `long` necessary since the max message size is [512 mb](http://redis.io/topics/protocol#resp-bulk-strings), we should be able to handle this with `int` right (and throw if out of range when parsing)?